opm/output/eclipse/EclipseReader.cpp

```diff
@@ -17,230 +17,163 @@
-        }
+    inline data::Wells restoreOPM_XWEL( ecl_file_type* file,
+                                        int num_wells,
+                                        int num_phases ) {
+        const char* keyword = "OPM_XWEL";
+
+        ecl_kw_type* xwel = ecl_file_iget_named_kw( file, keyword, 0 );
+        const double* xwel_data = ecl_kw_get_double_ptr(xwel);
+        const double* xwel_end  = xwel_data + ecl_kw_get_size( xwel );
+
+        const double* bhp_begin = xwel_data;
+        const double* bhp_end = bhp_begin + num_wells;
+        const double* temp_begin = bhp_end;
+        const double* temp_end = temp_begin + num_wells;
+        const double* wellrate_begin = temp_end;
+        const double* wellrate_end = wellrate_begin + (num_wells * num_phases);
+
+        const auto remaining = std::distance( wellrate_end, xwel_end );
+        const auto perf_elems = remaining / 2;
+
+        const double* perfpres_begin = wellrate_end;
+        const double* perfpres_end = perfpres_begin + perf_elems;
+        const double* perfrate_begin = perfpres_end;
+        const double* perfrate_end = perfrate_begin + perf_elems;
+
+        return { {},
+            { bhp_begin, bhp_end },
+            { temp_begin, temp_end },
+            { wellrate_begin, wellrate_end },
+            { perfpres_begin, perfpres_end },
+            { perfrate_begin, perfrate_end }
+        };
     }
-
-
-
-    void init_from_restart_file(EclipseStateConstPtr eclipse_state,
-                                int numcells,
-                                const PhaseUsage& phase_usage,
-                                SimulationDataContainer& simulator_state,
-                                WellState& wellstate) {
-
-        InitConfigConstPtr initConfig        = eclipse_state->getInitConfig();
-        IOConfigConstPtr ioConfig            = eclipse_state->getIOConfig();
-        int restart_step                     = initConfig->getRestartStep();
-        const std::string& restart_file_root = initConfig->getRestartRootName();
-        bool output                          = false;
-        const std::string& restart_file_name = ioConfig->getRestartFileName(restart_file_root, restart_step, output);
-
-        Opm::restoreSOLUTION(restart_file_name, restart_step, ioConfig->getUNIFIN(), eclipse_state, numcells, phase_usage, simulator_state);
-        Opm::restoreOPM_XWELKeyword(restart_file_name, restart_step, ioConfig->getUNIFIN(), wellstate);
+}
+
+std::pair< data::Solution, data::Wells >
+init_from_restart_file( const EclipseState& es, int numcells ) {
+
+    InitConfigConstPtr initConfig        = es.getInitConfig();
+    IOConfigConstPtr ioConfig            = es.getIOConfig();
+    int restart_step                     = initConfig->getRestartStep();
+    const std::string& restart_file_root = initConfig->getRestartRootName();
+    bool output                          = false;
+    const std::string& filename          = ioConfig->getRestartFileName(
```

Comment:
The `getRestartFileName()` has value return (should probably a have a better name ....); so I think the filename should be a value as well.